### PR TITLE
fix(extmarks): do not remove invalid marks from decor upon deletion

### LINKED ISSUE #27055: Neovim aborts when nvim_buf_del_extmark() is called on an invalid extmark
### Problem

Neovim aborts when nvim_buf_del_extmark() is called on an invalid extmark.

### Steps to reproduce

1. Run the code below. An inline virtual text is displayed on the 1st line.
```
call setline(1, ['11111', '22222', '33333'])
let ns = nvim_create_namespace('test')
let bn = bufnr('%')
call nvim_buf_set_extmark(bn, ns, 0, 5, #{virt_text: [['<-- ERROR!!!', 'Error']], virt_text_pos: 'inline', invalidate: v:true})
```
2. Delete the 1st line (:1d). The inline virtual text is hidden.
3. Show the ID of the extmark. `[[1, 0, 0]]` is echoed.
```
echo nvim_buf_get_extmarks(bn, ns, 0, -1, {})
```
4. Delete the hidden virtual text. Then, neovim suddenly aborts.
```
call nvim_buf_del_extmark(bn, ns, 1)
```
**Note:**
At step 3, try to show the details.
```
echo nvim_buf_get_extmarks(bn, ns, 0, -1, #{details: v:true})
```
echoes:
```
[1, 0, 0, {'virt_text_hide': v:false, 'priority': 4096, 'invalid': v:true, 'virt_text_repeat_linebreak': v:false, 'virt_text': [['<-- ERROR!!!', 'Error']], 'right_gravity': v:true, 'ns_id': 2, 'virt_text_pos': 'inline', 'invalidate': v:true}]
```
`'invalid': v:true` entry is present but this entry is not documented in help.

### Expected behavior

nvim_buf_del_extmark() can delete an invalid extmark and neovim does not abort.

### Neovim version (nvim -v)

NVIM v0.10.0-dev-2098+g92672a161

### Vim (not Nvim) behaves the same?

no

### Operating system/version

Windows 10

### Terminal name/version

n/a

### $TERM environment variable

n/a

### Installation

download from Github

### PR DESCRIPTION
**Pull request by luukvbaal**
**fix(extmarks): do not remove invalid marks from decor upon deletion**

Fix #27055

**docs(extmarks): add undocumented "details" array fields**

**refactor(extmarks): remove unused new pos from ExtmarkSavePos**

### CODE REVIEW

File: src/nvim/extmark.c [hunk 1]
```diff
@@ -394,14 +394,13 @@ void extmark_apply_undo(ExtmarkUndoObject undo_info, bool undo)
...
-        buf_put_decor(curbuf, mt_decor(mark), mark.pos.row, end.row);
+        if (extmark_setraw(curbuf, pos.mark, pos.old_row, pos.old_col) && pos.invalidated) {
+          MarkTreeIter itr[1] = { 0 };
+          MTKey mark = marktree_lookup(curbuf->b_marktree, pos.mark, itr);
+          mt_itr_rawkey(itr).flags &= (uint16_t) ~MT_FLAG_INVALID;
+          MTPos end = marktree_get_altpos(curbuf->b_marktree, mark, itr);
+          buf_put_decor(curbuf, mt_decor(mark), mark.pos.row, end.row);
+        }
       }
       // Redo
```
Comment (luukvbaal) [inline src/nvim/extmark.c:405]: @bfredl should this branch be removed or may it be used in the future? Currently `pos.row` always equals `-1`.

File: src/nvim/decoration.c [hunk 2]
```diff
@@ -212,8 +212,14 @@ void buf_put_decor_sh(buf_T *buf, DecorSignHighlight *sh, int row1, int row2)
   }
 }
 
-void buf_decor_remove(buf_T *buf, int row1, int row2, DecorInline decor, bool free)
+void buf_decor_remove(buf_T *buf, int row1, int row2, MTKey mark, bool free)
 {
+  DecorInline decor = mt_decor(mark);
+  if (mark.flags & MT_FLAG_INVALID) {
```
Comment (zeertzjq) [inline src/nvim/decoration.c:218]:
```suggestion
  if (mt_invalid(mark)) {
```

**Comment by luukvbaal** [comment]
Refactored for less churn.

File: test/functional/api/extmark_spec.lua [hunk 3]
```diff
@@ -1712,6 +1712,10 @@ describe('API/extmarks', function()
           aaa bbb ccc                         |*2
                                               |
     ]])
+    -- decor is not removed twice
```
Comment (luukvbaal) [inline test/functional/api/extmark_spec.lua:1715]: For what it's worth, this test will likely be ineffective after #27011 (as it removes the assert).

File: src/nvim/extmark.c (hunk 1 above)
Comment (bfredl) [reply src/nvim/extmark.c]: it's fine to delete it.

File: test/functional/api/extmark_spec.lua (hunk 3 above)
Comment (bfredl) [reply test/functional/api/extmark_spec.lua:1715]: to make things simple, #27011 doesn't care about invalidation at all. but once the basic version is done and merged, we could consider it as a follow-up.

Comment (luukvbaal) [reply test/functional/api/extmark_spec.lua:1715]: Sure. To be clear, I was just commenting on the effectiveness of the added test here. Not that the fact that `mt_meta_total()` doesn't take into account invalid marks will necessarily result in problems. The fact that it doesn't just means that this test would hit a different assert without the fix in this PR.

**Pull request merged**